preliz/distributions/asymmetric_laplace.py

```diff
@@ -0,0 +1,238 @@
+            self.params_support = ((eps, 1 - eps), (-np.inf, np.inf), (eps, np.inf))
+
+        self.kappa = kappa
+        self.mu = mu
+        self.b = b
+        if all_not_none(kappa, mu, b):
+            self._update(kappa, mu, b)
+
+    def _from_q(self, q):
+        kappa = (q / (1 - q)) ** 0.5
+        return kappa
+
+    def _to_q(self, kappa):
+        q = kappa**2 / (1 + kappa**2)
+        return q
+
+    def _update(self, kappa, mu, b):
+        self.kappa = np.float64(kappa)
+        self.mu = np.float64(mu)
+        self.b = np.float64(b)
+        self.q = self._to_q(self.kappa)
+
+        if self.param_names[0] == "kappa":
+            self.params = (self.kappa, self.mu, self.b)
+        elif self.param_names[0] == "q":
+            self.params = (self.q, self.mu, self.b)
+
+        self.is_frozen = True
+
+    def pdf(self, x):
+        """
+        Compute the probability density function (PDF) at a given point x.
+        """
+        x = np.asarray(x)
+        return np.exp(nb_logpdf(x, self.mu, self.b, self.kappa))
+
+    def cdf(self, x):
+        """
+        Compute the cumulative distribution function (CDF) at a given point x.
+        """
+        x = np.asarray(x)
+        return nb_cdf(x, self.mu, self.b, self.kappa)
+
+    def ppf(self, q):
+        """
+        Compute the percent point function (PPF) at a given probability q.
+        """
+        q = np.asarray(q)
+        return nb_ppf(q, self.mu, self.b, self.kappa)
+
+    def logpdf(self, x):
+        """
+        Compute the log probability density function (log PDF) at a given point x.
+        """
+        return nb_logpdf(x, self.mu, self.b, self.kappa)
+
+    def _neg_logpdf(self, x):
+        """
+        Compute the neg log_pdf sum for the array x.
+        """
+        return nb_neg_logpdf(x, self.mu, self.b, self.kappa)
+
+    def entropy(self):
+        return nb_entropy(self.b, self.kappa)
+
+    def median(self):
+        if self.kappa > 1:
+            return self.mu + self.kappa * self.b * np.log(
+                (1 + self.kappa**2) / (2 * self.kappa**2)
+            )
+        return self.mu - np.log((1 + self.kappa**2) / 2) / (self.kappa / self.b)
+
+    def mean(self):
+        kap_inv = 1 / self.kappa
+        kappa_dif = kap_inv - self.kappa
+        return kappa_dif * self.b + self.mu
+
+    def var(self):
+        kap_inv = 1 / self.kappa
+        return (kap_inv**2 + self.kappa**2) * self.b**2
+
+    def std(self):
+        return self.var() ** 0.5
+
+    def skewness(self):
+        return 2.0 * (1 - np.power(self.kappa, 6)) / np.power(1 + np.power(self.kappa, 4), 1.5)
+
+    def kurtosis(self):
+        return 6.0 * (1 + np.power(self.kappa, 8)) / np.power(1 + np.power(self.kappa, 4), 2)
+
+    def rvs(self, size=1, random_state=None):
+        random_state = np.random.default_rng(random_state)
+        random_samples = random_state.uniform(size=size)
+        # Returning ppf of random samples
+        return self.ppf(random_samples)
```

Comment:
something like this should work

@nb.njit()
def arvs(mu, b, kappa, size=1):
    u = np.random.uniform(-kappa, 1/kappa, size)
    s = np.sign(u)
    return mu - (1 / (1/b * s * kappa**s)) * np.log(1 - u * s * kappa**s)

Reply:
Almost similar I did something like this
```
@nb.vectorize(nopython=True)
def nb_rvs(U, mu, b, kappa):
    if -kappa < U < 1 / kappa:
        sgn = np.sign(U)
        return mu - (b / (sgn * kappa ** sgn)) * np.log(1 - U * sgn * kappa ** sgn)
    return nb_ppf(U, mu, b, kappa)
```
And call the nb_rvs in `rvs`

tried testing it out as well
```
asm = AsymmetricLaplace(kappa=2, mu=0, b=1)
asm1 = laplace_asymmetric(kappa=2, loc=0, scale=1)
print(asm.rvs(10, random_state=np.random.default_rng(1)))
print(asm1.rvs(10,random_state=np.random.default_rng(1)))

output
[-0.89327111  0.69780578  0.17006291  0.67982083  0.488635    0.93752564
  0.07454764  0.85296965 -0.75086494  0.02834773]
[-0.89327111  0.69780578 -3.42738124  0.67982083 -1.88429781 -1.2729362
  0.07454764 -1.34081961 -0.75086494 -6.73655692]
```

Reply:
You will need to do something like this, otherwise numba is going to complain about the random uniform or something like that


def rvs(mu, b, kappa, size=1, random_state=None):
    random_state = np.random.default_rng(random_state)
    u = np.random.uniform(-kappa, 1/kappa, size,)
    return nb_rvs(u, mu, b, kappa)

@nb.njit()
def nb_rvs(u, mu, b, kappa):
    s = np.sign(u)
    return mu - (1 / (1/b * s * kappa**s)) * np.log(1 - u * s * kappa**s)
    
    
* Set the seed following the test.
* Notice that you need to sample uniformly in the range [-kappa, 1/kappa] and transform all the samples. not sample U(0, 1) and transform the ones inside [-kappa, 1/kappa]

Reply:
Hmm...
I did that so that we could test our function with `scipy.stats`, since they just use `uniform(0, 1)` and then apply `ppf`.
Now since we will be changing the range of uniform, should I skip this test for `AsymmetricLaplace`

Reply:
I made a mistake in the previous snippet. This should give exactly the same result compared to scipy (up to machine precision)

def rvs(mu, b, kappa, size=1, random_state=None):
    rng = np.random.default_rng(random_state)
    u = rng.uniform(-kappa, 1/kappa, size,)
    return nb_rvs(u, mu, b, kappa)

@nb.njit()
def nb_rvs(u, mu, b, kappa):
    s = np.sign(u)
    return mu - (1 / (1/b * s * kappa**s)) * np.log(1 - u * s * kappa**s)

Reply:
@aloctavodia we will be using ` u = rng.uniform(-kappa, 1/kappa, size,)`, whereas scipy uses `rng.uniform(size=size)` and then apply ppf.
When we set the range to [-kappa, 1/kappa] we will have negative values as well which will be calculated in our case whereas that is not the case in the `scipy.stats` one.

Reply:
But we are not going to use the ppf (that is only defined for the interval [0, 1]). We are going to transform those uniform values using the nb_rvs function.

Reply:
Yeah I know, I am asking about the `test_scipy.py` tests, there we need to skip the tests for `rvs`

Reply:
No need to skip it, I already make the comparison and is essentially the same

Reply:
Thanks, I realized my mistake...
I realized that we are trying to map our `[-kappa, 1/kappa]` uniform distribution to scipy's `[0,1]`.

I was trying to equate the answer of each U value and that's where I was getting it wrong..